### PR TITLE
Ctoken payable wrapper

### PR DESCRIPTION
**Pull request by haythem96**
# Ctoken payable wrapper

## High Level Description

Combining cToken on ERC-20s (non cETH) with Payable proxy for ETH/WETH

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [x] Is your code up to date with the spec? 
- [x] Have you added your tests to the testing doc?

### CODE REVIEW

File: contracts/external/proxies/CERC20Proxy.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                    (msg.sender == action.owner) || (controller.isOperator(action.owner, msg.sender)),
+                    "CERC20Proxy: msg.sender is not owner or operator "
+                );
+            }
+
+            // overwrite the deposit amount by the exact amount minted
+            if (action.actionType == Actions.ActionType.DepositCollateral && action.amount == 0) {
+                // should we only update if the action.second address == address(this)?
+                _actions[i].amount = cTokenBalance;
+            }
```
Comment (antoncoding) [inline contracts/external/proxies/CERC20Proxy.sol:77]:
```suggestion
            if (action.actionType == Actions.ActionType.DepositCollateral && action.amount == 0) {
                _actions[i].amount = cTokenBalance;
            }
```
Don't need to add that condition, if user has cToken just not use action.amount = 0

Comment (antoncoding) [reply contracts/external/proxies/CERC20Proxy.sol:77]: Can remove the comment

File: contracts/external/proxies/PayableCERC20.sol [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                );
+            }
+
+            // overwrite the deposit amount by the exact amount minted
+            if (
+                action.actionType == Actions.ActionType.DepositCollateral &&
+                action.amount == 0 &&
+                action.asset == address(cToken)
+            ) {
+                // should we only update if the action.second address == address(this)?
```
Comment (antoncoding) [inline contracts/external/proxies/PayableCERC20.sol:103]:
```suggestion
```

File: contracts/external/proxies/PayableCERC20.sol [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                _actions[i].amount = cTokenBalance;
+            }
+            // check if Call action exist, approve WETH to Callee
+            if (action.actionType == Actions.ActionType.Call) {
+                ERC20Interface(address(weth)).safeIncreaseAllowance(
+                    action.secondAddress,
+                    weth.balanceOf(address(this))
```
Comment (antoncoding) [inline contracts/external/proxies/PayableCERC20.sol:110]:
```suggestion
                    msg.value
```
I think using msg.value is equivalent to `weth.balanceOf(address(this))` in all cases, and it saved us some gas

**Comment by antoncoding** [comment]
Also another question is, do we need the cETH wrapper? I thought our decision was to only wrap USDC for puts

File: contracts/external/proxies/CERC20Proxy.sol (hunk 1 above)
Comment (aleone) [reply contracts/external/proxies/CERC20Proxy.sol:77]: Yeah that was internal monologue that I forgot to delete :)

**Review by aparnakr — APPROVED** [review]

**Comment by haythem96** [comment]
Should we remove `external/proxies/CERC20Proxy.sol` and `external/proxies/CETHProxy.sol` or just ignore them for now ?

**Comment by antoncoding** [comment]
> Should we remove `external/proxies/CERC20Proxy.sol` and `external/proxies/CETHProxy.sol` or just ignore them for now ?


We should remove them

**Comment by antoncoding** [comment]
One thing that'll be helpful for this wrapper to add is an additional input for the `operate` function that enable this contract to call oToken.approve(). This way we can let the wrapper has allowance for marginPool, so after we bought some long oToken, we can directly deposit it from the wrapper to the marginPool.

**Review by antoncoding — APPROVED** [review]
Plz fix the e2e project id before merging

**Comment by haythem96** [comment]
> Plz fix the e2e project id before merging

Fixed.

**Comment by haythem96** [comment]
Closing this for now.

**Pull request closed without merging**